src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,283 @@
+   *
+   * canResume contains the logic that will decide if the client will retry with
+   * another request when it receives an error. This logic is passed along to
+   * google-gax and used by google-gax to decide if the client should retry
+   * a request when google-gax receives an error. If canResume returns true then
+   * the client will retry with another request computed by getResumeRequest. If
+   * canResume request returns false then the error will bubble up from gax to
+   * the handwritten layer.
+   *
+   * @param {GoogleError} [error] The error that Google Gax receives.
+   * @return {boolean} True if the client will retry
+   */
+  canResume(error: GoogleError): boolean {
+    // First update the row keys and the row ranges based on the last row key.
+    this.updateKeysAndRanges();
+    if (error.statusDetails === 'RetryInfo') {
+      return true;
+    }
+    // If all the row keys and ranges are read, end the stream
+    // and do not retry.
+    if (this.rowKeys.length === 0 && this.ranges.length === 0) {
+      return false;
+    }
+    // If there was a row limit in the original request and
+    // we've already read all the rows and met/exceeded that limit, end the
+    // stream and do not retry.
+    if (this.hasLimit && this.rowsLimit === this.rowsRead) {
+      return false;
+    }
+    const retryCodesUsed = this.retryCodes
+      ? this.retryCodes
+      : arrify(RETRYABLE_STATUS_CODES);
+    if (
+      error.code &&
+      (retryCodesUsed.includes(error.code) || isRstStreamError(error))
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Creates a RetryOptions object that can be used by google-gax.
+   *
+   * This class contains the business logic to specify retry behaviour of
+   * readrows requests and this function packages that logic into a RetryOptions
+   * object that google-gax expects.
+   *
+   * @param {CallOptions} [gaxOpts] The call options that will be used to
+   * specify retry behaviour.
+   * @return {RetryOptions} A RetryOptions object that google-gax expects that
+   * can determine retry behaviour.
+   *
+   */
+  toRetryOptions(gaxOpts: CallOptions): RetryOptions {
+    // On individual calls, the user can override any of the default
+    // retry options. Overrides can be done on the retryCodes, backoffSettings,
+    // shouldRetryFn or getResumptionRequestFn.
+    const canResume = (error: GoogleError) => {
+      return this.canResume(error);
+    };
+    const getResumeRequest = () => {
+      return this.getResumeRequest() as RequestType;
+    };
+    // In RetryOptions, the retryCodes are ignored if a shouldRetryFn is provided.
+    // The 3rd parameter, the shouldRetryFn will determine if the client should retry.
+    return new RetryOptions(
+      [],
+      gaxOpts?.retry?.backoffSettings || DEFAULT_BACKOFF_SETTINGS,
+      gaxOpts?.retry?.shouldRetryFn || canResume,
+      gaxOpts?.retry?.getResumptionRequestFn || getResumeRequest
```

Comment:
Yes, you are right. If `shouldRetryFn` is provided then it will override `canResume` and if `getResumptionRequestFn` is provided then it will override `getResumeRequest`. The pros of this is that it gives the user complete control of how they want retries to work. The cons are the cons that you mentioned. `should use a different set of retry codes` is now a test case that covers users providing custom retry codes. Do you think we should not allow the user to override `canResume` and `getResumeRequest`? I wonder what @leahecole thinks.

Reply:
Personally, I would recommend making them non-customizable (i.e. raising an exception if the user sets these themselves). It would be easy enough to make them customizable in the future if there's demand, but once it's part of the public surface, we have to maintain it going forward. And I can imagine a bunch of weird edge cases coming up, that I'm not confident are covered by tests

But I'll leave that up to you and the product team

Reply:
I think this is a good point. We can always add customizability later.

Reply:
+1, customizability can always happen later. I think that client library teams should be able to customize them (which is what we are doing with this PR) but giving the user too much control here could lead to a lot of issues and definitely could be hard to maintain. Good point, @daniel-sanche